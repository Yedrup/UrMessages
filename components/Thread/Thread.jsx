import PropTypes from 'prop-types';
import MessageItem from '../Message/Message';
import MessagesList from '../MessagesList/MessagesList';
import CreateMessage from '../CreateMessage/CreateMessage';
import styled from 'styled-components';

const InitialMessage = styled.div`
  margin: 2rem 0;
  background: black;
  color: white;
  article {
    margin: auto;
    background: black;
    color: #aaeeee;
  }
`;

const Thread = ({ parentMessage, thread }) => {
  let isThreadMessages = !!thread?.messages?.length;
  return (
    <div>
      <h2>Thread of Message {parentMessage.id} </h2>
      <InitialMessage>
        <MessageItem message={{ ...parentMessage, isThreadAllowed: false }} />
      </InitialMessage>
      {isThreadMessages && <MessagesList messages={thread.messages} />}
      {!isThreadMessages && (
        <h3 data-testid="start-thread-message">Start a thread ! ğŸ‘‡</h3>
      )}
      <CreateMessage
        ctx={{
          isPublic: parentMessage.isPublic,
          parentId: parentMessage.id,
          id: thread?.id,
        }}
      />
    </div>
  );
};

Thread.propTypes = {
  parentMessage: PropTypes.object,
  thread: PropTypes.object,
};

export default Thread;
