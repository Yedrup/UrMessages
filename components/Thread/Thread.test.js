import { render, cleanup } from '@testing-library/react';
import PropTypes from 'prop-types';
import DataProvider from '../../contexts/DataContext';
import UIProvider from '../../contexts/UIContext';
import Thread from './Thread';

import {
  fakeThread,
  fakeThreadParent,
  fakeThreadWithoutMessages,
} from '../../lib/testUtils';

const ContextWrapper = ({ children }) => (
  <DataProvider>
    <UIProvider>{children}</UIProvider>
  </DataProvider>
);

ContextWrapper.propTypes = {
  children: PropTypes.object,
};

describe('<Thread/>', () => {
  afterEach(cleanup);
  it('properly renders the component - snapshot', () => {
    const { asFragment } = render(
      <ContextWrapper>
        <Thread parentMessage={fakeThreadParent} thread={fakeThread} />
      </ContextWrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the MessageList', () => {
    const { queryByTestId } = render(
      <ContextWrapper>
        <Thread parentMessage={fakeThreadParent} thread={fakeThread} />
      </ContextWrapper>
    );
    expect(queryByTestId('thread-message-list')).toBeInTheDocument();
  });

  it('should not render the MessageList cause there is no messages', () => {
    const { queryByTestId } = render(
      <ContextWrapper>
        <Thread
          parentMessage={fakeThreadParent}
          thread={fakeThreadWithoutMessages}
        />
      </ContextWrapper>
    );
    expect(queryByTestId('thread-message-list')).not.toBeInTheDocument();
  });

  it('should render a message to start the thread cause there is no messages', () => {
    const { queryByTestId } = render(
      <ContextWrapper>
        <Thread
          parentMessage={fakeThreadParent}
          thread={fakeThreadWithoutMessages}
        />
      </ContextWrapper>
    );
    expect(queryByTestId('start-thread-message')).toBeInTheDocument();
  });

  it('should not render the MessageList cause there is no messages', () => {
    const { queryByTestId } = render(
      <ContextWrapper>
        <Thread
          parentMessage={fakeThreadParent}
          thread={fakeThreadWithoutMessages}
        />
      </ContextWrapper>
    );
    expect(queryByTestId('thread-message-list')).not.toBeInTheDocument();
  });
});
