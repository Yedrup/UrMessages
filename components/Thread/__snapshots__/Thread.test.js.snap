// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Thread/> properly renders the component - snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Thread of Message c4b1e9e2-0ca6-4a84-b6e5-538241f1b447 
    </h2>
    <div
      class="Thread__InitialMessage-sc-1ygzf59-0 gANVOl"
    >
      <article
        class="Message__StyledMessageItem-uraoec-0 dDlRBV"
        data-testid="message-item"
      >
        <header>
          <h3>
            Lorem ipsum dolor sit amet consectetur 
            <span>
              ðŸ”’
            </span>
          </h3>
        </header>
        <section>
          Doloremque tenetur dolore iure nisi quidem laudantium, hic earum necessitatibus
          <p>
            c4b1e9e2-0ca6-4a84-b6e5-538241f1b447
          </p>
        </section>
        <footer>
          <p>
            Sunday, August 2, 2020, 1:47:56 PM
          </p>
        </footer>
      </article>
    </div>
    <div
      class="MessagesList__ListOfItemsStyled-epu0q5-0 dnvssR"
      data-testid="thread-message-list"
    >
      <article
        class="Message__StyledMessageItem-uraoec-0 dDlRBV"
        data-testid="message-item"
      >
        <header>
          <h3>
            Delectus itaque dolores 
            <span>
              ðŸ”’
            </span>
          </h3>
        </header>
        <section>
          odio illo possimus veniam qui ratione nihil
          <p>
            934bae862-bbb7-4b9d-83fe-555c8734be4e
          </p>
        </section>
        <footer>
          <p>
            Sunday, August 2, 2020, 9:35:52 PM
          </p>
        </footer>
      </article>
    </div>
    <form
      class="FormStyled-j8ogoe-0 iIBCUw"
      data-test="form"
      id="test"
    >
      <fieldset
        aria-busy="false"
      >
        <label
          for="title"
        >
          Title
          <input
            id="title"
            name="title"
            placeholder="Title"
            required=""
            type="text"
            value=""
          />
        </label>
        <label
          for="content"
        >
          Message
          <textarea
            id="content"
            name="content"
            placeholder="Enter A Message"
            required=""
          />
        </label>
        <button
          class="c-button"
          data-testid="button-submit-message"
          type="submit"
        >
          Submit
        </button>
      </fieldset>
    </form>
  </div>
</DocumentFragment>
`;
